armbian/build
# Tinkerboard camera support

Finally.. the camera works on the tinkerboard.. This patch series implements both RPi cameras into rockchip-default and the missing devicetree overlay infrastructure (we support it on rockchip-next, but not yet on the 4.4 kernel).

To test you need:
- obviously a tinkerboard
- ov5647 (rpi cam v1.3) or imx219 (rpi cam v2.1)
- @teacupx multimedia script
- gstreamer1.0-tools (sudo apt get install)
- (currently) https://github.com/rockchip-linux/rk-rootfs-build/blob/88a256042abbd2d44fe7fe3250c2adf5545caa39/overlay-firmware/etc/cam_iq/OV5647.xml or https://github.com/rockchip-linux/rk-rootfs-build/blob/88a256042abbd2d44fe7fe3250c2adf5545caa39/overlay-firmware/etc/cam_iq/imx219.xml (honestly, I didn't care much about the userspace side of this... our multimedia wizards will be much better in this).

OV5647:
`gst-launch-1.0 rkcamsrc num-buffers=10 io-mode=4 isp-mode=3A tuning-xml-path=ov5647.xml ! video/x-raw,format=NV12,width=1920,height=1080 ! jpegenc ! multifilesink location=image.jpg`
![image](https://user-images.githubusercontent.com/21057365/61965675-08130280-afd1-11e9-9561-83bcdcb4706f.jpg)

and for videos:
`gst-launch-1.0 v4l2src ! video/x-raw,format=NV12,width=1920,height=1080,framerate=30/1 ! queue ! mpph264enc ! queue ! h264parse ! mpegtsmux ! filesink location=/home/opi/2k.ts`
https://youtu.be/zSM6ozBqrMo

switching to imx219 (dt overlay):
```
opi@tinkerboard:~$ dmesg | grep imx
[    2.531304] imx219 2-0010: Model ID 0x0219, Lot ID 0x5c6b63, Chip ID 0x01c2
```
Notice: the ov5647 doesn't really show up in dmesg, and the camera light isn't an indicator that it works, I had it recording with and without red LED being on (it comes on for a short period during boot in all cases, but after.. no idea what and how it's triggered)
the picture:
`gst-launch-1.0 rkcamsrc num-buffers=10 io-mode=4 isp-mode=3A tuning-xml-path=imx219.xml ! video/x-raw,format=NV12,width=1920,height=1080 ! jpegenc ! multifilesink location=image.jpg`
![image](https://user-images.githubusercontent.com/21057365/61971477-79f24880-afdf-11e9-83e8-c9a6d1280d46.jpg)

the video:
`gst-launch-1.0 v4l2src ! video/x-raw,format=NV12,width=1920,height=1080,framerate=30/1 ! queue ! mpph264enc ! queue ! h264parse ! mpegtsmux ! filesink location=/home/opi/2k.ts`
https://youtu.be/8sqWxgAICMg

currently tested:
- both cameras
- at least the overlay for he debug UART works (otherwise UART would go silent during boot)

**not tested**:
- all other overlays, I don't see a reason that something can go wrong here but who knows..

known 'bugs' & unknowns:
- calling the camera without proper dt as it was before commit 21c3726  ends in a kernelcrash :dancer: 
- imx is dark & greenish (ov5647 as well, but it's looking better), therefore ov5647 is currently the default camera (it's a won't fix, cause I've no idea how this things working and no motivation to dive into it) others have (had?) pink camera output so... **Dark green is the new pink!** https://github.com/raspberrypi/firmware/issues/1167
- there's new stuff out to support cameras in rockchips repository, maybe stuff from @teacupx multimedia script has to be updated. 
- userspace access to the camera is far from done. It works with a bit of handwork, stuff is missing here and there but well, it is what it is. 
- I didn't check preview functions of the camera, still missing a HDMI scree for such tests.. :rofl:
- probably those xml files should be loaded for better video quality, but I think it's a nice task for someone to contribute.. :smile_cat: I forgot most of the other commands used during debug.. I thought I had once a better video for the IMX219 cam.. 
- the patches are a bit hacky (especially the dt overlay stuff) but it should be maintainable.
